molgenis-security/src/main/java/org/molgenis/security/permission/PermissionSystemServiceImpl.java

```diff
@@ -20,79 +18,103 @@
 import org.springframework.security.core.context.SecurityContextHolder;
 import org.springframework.stereotype.Component;
 
+import java.util.Collection;
 import java.util.List;
 import java.util.stream.Stream;
 
+import static java.util.Collections.singleton;
 import static java.util.Objects.requireNonNull;
 import static java.util.stream.Collectors.toList;
 import static org.molgenis.auth.UserAuthorityMetaData.USER_AUTHORITY;
-import static org.molgenis.auth.UserMetaData.USER;
+import static org.molgenis.security.core.Permission.WRITEMETA;
 import static org.molgenis.security.core.runas.RunAsSystemProxy.runAsSystem;
+import static org.molgenis.security.core.utils.SecurityUtils.AUTHORITY_ENTITY_PREFIX;
 
 @Component
 public class PermissionSystemServiceImpl implements PermissionSystemService
 {
-	private final DataService dataService;
+	private final UserService userService;
 	private final UserAuthorityFactory userAuthorityFactory;
+	private final RoleHierarchy roleHierarchy;
+	private final DataService dataService;
 
 	@Autowired
-	public PermissionSystemServiceImpl(DataService dataService, UserAuthorityFactory userAuthorityFactory)
+	public PermissionSystemServiceImpl(UserService userService, UserAuthorityFactory userAuthorityFactory,
+			RoleHierarchy roleHierarchy, DataService dataService)
 	{
-		this.dataService = requireNonNull(dataService);
+		this.userService = requireNonNull(userService);
 		this.userAuthorityFactory = requireNonNull(userAuthorityFactory);
+		this.roleHierarchy = requireNonNull(roleHierarchy);
+		this.dataService = requireNonNull(dataService);
+	}
+
+	@Override
+	public void giveUserEntityPermissions(EntityType entityType)
```

Comment:
maybe rename to something like 'giveUserWriteMetaPermissions' to hint at the fact Write_Meta permissions are given.